### ISSUE TITLE
Postgres sync not respecting SSH tunneling

### ISSUE THREAD

**Issue by mrdeds** [metabase/metabase#8396]
### Bugs

- Browser and the version: Chrome Versión 68.0.3440.106 (Build oficial) (64 bits)
- Operating system: OS X 10.13.5
- Databases: Redshift
- Metabase version:  0.30.1
- Metabase hosting environment: 5 Kubernets pods with 2 GB CPUs and 2 GB memory each
- Metabase internal database: Postgres

 *Repeatable steps to reproduce the issue*
Open a new personalized query, select a database with more than one schema.

Here's what I see sometimes:
<img width="1440" alt="captura de pantalla 2018-08-24 a la s 16 21 20" src="https://user-images.githubusercontent.com/9302082/44608792-be1b0e80-a7ba-11e8-89c1-6d1256a3a15f.png">

And sometimes all schemas are shown, I should see also a **"Public"** schema.

It allows me to use a raw query to return results but it is not pulling the **Public** schema in the GUI.

Here's how it should look:
<img width="1440" alt="captura de pantalla 2018-08-24 a la s 16 54 21" src="https://user-images.githubusercontent.com/9302082/44609698-641c4800-a7be-11e8-9b9e-27f378913aa8.png">

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by jhtravis** [comment]
I also am experiencing this issue.  Metabase version 0.30.1, also a Postgres database.  It pulls the database table names and allows me to use a raw query to return results but it is not pulling column names at all.

**Comment by jornh** [comment]
@mrdeds @jhtravis when you experience the missing schema issue:
- do you see any errors in the browser console? 
- Is this a new thing in 0.30.1 or seen in any previous versions (what did you upgrade from?)

@jhtravis also experienced intermittently for you? Maybe a separate issue, but can you expand on the “not pulling column names at all”. Screenshot? Does admin database sync etc. change anything? Logs from server on sync/browser on missing columns showing anything abnormal?

**Comment by mrdeds** [comment]
It’s been happening a lot in other versions too

**Comment by mrdeds** [comment]
Yeap, in multiple pods too and single VM previously

**Comment by jhtravis** [comment]
Attached is a screenshot of the tables being found but the columns not being retrieved.  Some of these tables do have PostGIS geometry columns.  But I have successfully pulled these columns before.  

I have re-scanned fields with no success.

![missing_columns](https://user-images.githubusercontent.com/4077468/44670306-55b57280-a9f0-11e8-92ba-88401dbc0638.PNG)

There are error logs, they are attached here:
Aug 27 16:00:01 ERROR metabase.sync.util :: Error running sync step: org.postgresql.util.PSQLException cannot be cast to java.util.Map$Entry
("sync.sync_metadata.fields$fn__41470$sync_fields_BANG___41475$fn__41476.invoke(fields.clj:367)"
"sync.sync_metadata.fields$fn__41470$sync_fields_BANG___41475.invoke(fields.clj:363)"
"sync.util$fn__35635$run_step_with_metadata__35640$fn__35644$fn__35646.invoke(util.clj:350)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invokeStatic(util.clj:98)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invoke(util.clj:92)"
"sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:115)"
"sync.util$with_start_and_finish_debug_logging.invoke(util.clj:112)"
"sync.util$fn__35635$run_step_with_metadata__35640$fn__35644.invoke(util.clj:347)"
"sync.util$fn__35635$run_step_with_metadata__35640.invoke(util.clj:342)"
"sync.util$fn__35719$run_sync_operation__35724$fn__35725$fn__35726.invoke(util.clj:397)"
"sync.util$fn__35719$run_sync_operation__35724$fn__35725.invoke(util.clj:397)"
"sync.util$fn__35719$run_sync_operation__35724.invoke(util.clj:391)"
"sync.sync_metadata$fn__42043$sync_db_metadata_BANG___42048$fn__42049$fn__42050.invoke(sync_metadata.clj:51)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:143)"
"sync.util$do_with_error_handling.invoke(util.clj:138)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:141)"
"sync.util$do_with_error_handling.invoke(util.clj:138)"
"driver$fn__28315.invokeStatic(driver.clj:288)"
"driver$fn__28315.invoke(driver.clj:288)"
"driver$fn__28250$G__28057__28259.invoke(driver.clj:104)"
"sync.util$sync_in_context$fn__35541.invoke(util.clj:134)"
"sync.util$with_db_logging_disabled$fn__35538.invoke(util.clj:125)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invokeStatic(util.clj:98)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invoke(util.clj:92)"
"sync.util$with_start_and_finish_logging$fn__35527.invoke(util.clj:110)"
"sync.util$with_sync_events$fn__35522.invoke(util.clj:84)"
"sync.util$with_duplicate_ops_prevented$fn__35513.invoke(util.clj:63)"
"sync.util$do_sync_operation.invokeStatic(util.clj:162)"
"sync.util$do_sync_operation.invoke(util.clj:159)"
"sync.sync_metadata$fn__42043$sync_db_metadata_BANG___42048$fn__42049.invoke(sync_metadata.clj:50)"
"sync.sync_metadata$fn__42043$sync_db_metadata_BANG___42048.invoke(sync_metadata.clj:47)"
"task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:36)")

**Comment by jhtravis** [comment]
I executed a rescan and see that it is not connecting to the individual tables.   Long log attached in text form
[logs.txt](https://github.com/metabase/metabase/files/2324706/logs.txt)

**Comment by jornh** [comment]
@sbelak I'm (wildly 😝) guessing that rescan log in https://github.com/metabase/metabase/issues/8396#issuecomment-416279447 seems like something that could give you a clue?

**Comment by jhtravis** [comment]
Hi all,  I was able to get the schema to sync on my docker image once I moved the network to be host.  I was trying to ssh into another docker image on the same server which was working on the connection part of Metabase but not on the schema read.  Moving the network to host resolved this.  Just wanted to let you know I am not blocked by this now but it still may be good to test schema refreshes to DB's that use an SSH connection.

**Comment by danwwilson** [comment]
As a follow up to my previous comment, if I drop the SSH connection and connect directly to the DB it does refresh the schema fine.

**Comment by jhtravis** [comment]
I am relooking at this same problem and realize that for my instance this bug essentially removes the Metabase SSH functionality.  Though I found a workaround in one instance this may be a breaking bug for another use case.  I welcome feedback if you are able to return schemas for databases through an SSH connection.

**Comment by minhtrietbkit** [comment]
We ran into this also.

We have set up ssh tunneling, only to realize that metabase has this bug. Now if we want to allow direct connect, we have to restart DB, which is not an option...

That makes metabase unusable right now for us.

Does anybody know of a workaround?

**Comment by seocam** [comment]
Basically the problem happens because the sync code currently don't use the tunnel configuration? If that's the case isn't this an easy-picking? :)

**Comment by mazameli** [comment]
A fair number of folks seem to be running into this, so bumping up priority.

**Comment by patrickrmaia** [comment]
We're also facing this issue. 

Our public schema has 41 tables (and user created views):

```
SELECT COUNT(*)
FROM information_schema.tables
WHERE table_schema = 'public'

count
41
```

But Metabase can only find 35 tables as stated in the `classify-tables` log:

`85874	classify-tables	2	2019-11-29 14:25:31.509	2019-11-29 14:25:31.548	39	"{**""total-tables"":35**,""tables-classified"":0}"`

**Comment by flamber** [comment]
@patrickrmaia Since this issue seems to be about sync'ing over SSH, I'm guessing you might have a different problem. There has been a lot of database connection fixes lately, so I would recommend that you try the latest 0.33.6 (remember to backup before upgrade!) and then run a forced scan (Admin > Databases) and check the log (Admin > Troubleshooting > Logs).

**Comment by opsydev** [comment]
This bug has been open for almost two years now and marked as P1. Is this still being considered P1?

**Comment by flamber** [comment]
@opsydev There's currently 47 other P1s, and there's multiple issues connected throughout using the integrated SSH functionality, which has made this (and the other related issues) more difficult to fix than expected. The workaround is to use an external tunnel:
https://www.metabase.com/docs/latest/administration-guide/01-managing-databases.html#what-if-the-built-in-ssh-tunnels-dont-fit-my-needs

**Comment by opsydev** [comment]
@flamber I totally understand. I am in the process of setting up metabase for my organization. I am trying to gauge when this bug would be fixed as it affects how I go about setting things up. If this is going to take a lot longer to fix that's fine. Seeing a P1 bug open for two years doesn't provide any indication about where in the Metabase teams priority list it actually is. 

If for the forseeable future it's not on the Radar then I'm happy to look at workaround, it would just mean it's more  permanent than just a temporary workaround. 

As I mention in the linked bug the workaround is not really manageable when there  are multiple servers and metabase is running on docker. Either I have would have to manage a jump box or in my case it's easier to open the databases directly to metabase.

**Comment by camsaul** [comment]
More information about this issue in #9278

**Comment by camsaul** [comment]
This actually looks like a really easy fix. 

The issue is [this code](https://github.com/metabase/metabase/blob/master/src/metabase/driver/postgres.clj#L97-L104) in the Postgres driver:

```clj
(defn- enum-types [driver database]
  (set
   (map (comp keyword :typname)
        (jdbc/query (sql-jdbc.conn/connection-details->spec driver (:details database))
                    [(str "SELECT DISTINCT t.typname "
                          "FROM pg_enum e "
                          "LEFT JOIN pg_type t "
                          "  ON t.oid = e.enumtypid")]))))
```

1.  `enum-types` needs to use the SSH tunnel if one exists.

2.  Even if `enum-types` fails I do not think it should cause sync to fail entirely.

We should also audit other driver code and make sure we don't have other calls to `jdbc/query` that don't respect SSH tunnels and hopefully come up with a general utility function to use tunnels automatically when appropriate. It would be nice to make this robust and implement #10081 as well but that can be done separately.

I think writing a test for this will be a little tricky

**Comment by jaredtkatz** [comment]
@camsaul any updates on this?

**Comment by dpsutton** [comment]
I might not have linked the PR to the issue correctly but this is fixed and in a release candidate: https://github.com/metabase/metabase/pull/13391.

Release information: https://github.com/metabase/metabase/releases/tag/v0.37.0-rc2

**Comment by mtorcutt** [comment]
This issue is now resolved with `v0.37.0-rc2` but the connection drops after a short period of time. I get an error of `Connection to localhost:37235 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.` As soon as I reset the DB (by resetting a value in the DB form), I have no issues and it reconnects for a short period of time again.

```
{
  "browser-info": {
    "language": "en",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.80 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.7+10",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.7",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.7+10",
    "os.name": "Linux",
    "os.version": "4.14.198-152.320.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "elastic-beanstalk",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.6"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-10-07",
      "tag": "v0.37.0-rc2",
      "branch": "master",
      "hash": "1f52afa"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Comment by flamber** [comment]
@mtorcutt There's a couple of related issues about tunnels timeout: #3941 #6114 #6173 #10081 #13399
